modified gcc-bridge to support basic c++ compilation

hwvenancio opened the pull request:
modified Builtins: added __finite, used by gcc-bridge-example
modified GccBridgeMojo: added file scanner that search in subfolders
modified Gcc: compiles .cpp files (uses g++-4.6)

added SymbolLibrary, SymbolFunction and SymbolMethod, used to configure GimpleCompiler+GlobalSymbolTable
added CppSymbolLibrary: defines c++ symbols (currently only operator new, __comp_ctor and __comp_dtor)
added CtorCallGenerator: generates code for __comp_ctor, not implemented yet

modified plugin.c: dumps all edges of basic blocks, refactored constants names in dump_statements
added GimpleEdge: represents a jump between GimpleBasicBlock
modified ControlFlowGraph: uses GimpleEdge instead of GimpleGoto to add edges between basic blocks

added simple c++ test (CppClassTest compiles cppclass.cpp and return new instance through method call)

akbertram commented:
@bddbot add to whitelist please!

akbertram commented on tools/gcc-bridge/maven-plugin/pom.xml:
```diff
@@ -31,6 +31,17 @@
       <scope>provided</scope>
     </dependency>
     <dependency>
+      <groupId>org.codehaus.plexus</groupId>
```
It looks like this addition is failing the build: 
https://jenkins.bedatadriven.com/job/Renjin/job/renjin-pull-requests/32/console

`plexusCompilerVersion` is not defined.

akbertram commented on tools/gcc-bridge/compiler/src/main/java/org/renjin/gcc/Gcc.java:
```diff
@@ -121,9 +122,12 @@ public GimpleCompilationUnit compileToGimple(File source, String... compilerFlag
+  private String callGcc(List<String> arguments, String extension) throws IOException {
     List<String> command = Lists.newArrayList();
-    command.add("gcc-4.6");
+    if("cpp".equals(extension))
```
This shouldn't be necessary - the `gcc` binary can be used for c, fortran, c++, etc -- it will send to the correct frontend.

hwvenancio replied:
Nice, I didn't knew that. My "natural language" is Java, not C++, but I'm really interested in converting some libraries to Java (mainly bullet physics engine).

akbertram replied:
Excellent! Would love to see gcc-bridge find use outside of renjin so happy to help in any way to get this working for other libraries.

akbertram commented:
@hwvenancio Thanks picking this up!

hwvenancio commented:
It didn't pass integration tests. I don't have R installed on this machine to test everything, normally I just run mvn clean install on renjin/tools/gcc-bridge/. I'll have a closer look later.

akbertram commented:
Hold on a sec, the build is failing on the master branch. @psolaimani is fixing now...

akbertram commented:
@bddbot retest this please

psolaimani commented:
Sorry for the inconvenience. I had moved over 100 tests from failingMicroTests to MicroTest since they are passing now. However, I accidentally moved a failing test (micro1420) as well.

hwvenancio commented:
Nice! Looks like I made my first contribution
@psolaimani no problem :)

akbertram commented:
Excellent!
Do you want to continue pushing changes to this PR or would you prefer to merge what is complete into the master branch now?

If you'd like to merge, then it would be great if you could address my comment on the gcc binary and remove any commented-out code. 

Thanks again!

hwvenancio commented:
I reverted those changes in Gcc.java and cleaned-up commented and duplicated code. I think it's ready to merge.
Now I need to learn a lot about bytecode generation, then maybe I can generate classes with multiple constructors and methods.

akbertram commented:
Have merged it! 
I've reformatted most of your sources to match the rest of the project's style, and have add a checkstyle plugin so that you get earlier feedback in the future.

As for multiple constructors, I suspect that GCC simply mangles the names so that you don't end up with just a series of normal functions. I believe the next challenge will be when you add  classes that require virtual dispatch. For example:
https://github.com/bedatadriven/renjin/blob/master/tools/gcc-bridge/compiler/src/test/java/org/renjin/gcc/GimpleCompilerTest.java#L480-480

I'll close this PR and we can continue on Issue #119

hwvenancio commented:
Maybe __comp_ctor is some internal function that calls the actual constructor method? I'll play around and see what gimple it generates.

As far as I know, when using the "new" operator, first c++ allocates some memory, then it calls the constructor, as seen with this test code:

```c++
void* operator new(size_t size) {
	cout << "global new" << endl;
	return malloc(size);
}
void operator delete(void* mem) {
	cout << "global delete" << endl;
	free(mem);
}
class Test1 {
public:
	Test1() {
		cout << "Test1 constructor" << endl;
	};
	virtual ~Test1() {
		cout << "Test1 destructor" << endl;
	}
};
int main(int argc, char* argv[]) {
	Test1* test1 = new Test1;
	delete test1;
}
```
Results in:
```
global new
Test1 constructor
Test1 destructor
global delete
```
I don't know if this is basic knowledge for C++ programmers, but as I said, I work mostly with Java.
